app/controllers/fhir/r4/api_controller.rb

```diff
@@ -342,12 +335,100 @@ def create
     else
       status_not_found && return
     end
     status_created(resource.as_fhir) && return
   rescue JSON::ParserError
     status_bad_request(['Invalid JSON in request body'])
+  rescue SaveError
+    nil
+  end
+
+  # Create a set of resources as an atomic action
+  #
+  # Supports (writing): Patient, Observation
+  #
+  # POST /fhir/r4
+  def transaction
+    status_unsupported_media_type && return unless content_type_header?('application/fhir+json')
+
+    # Must have both Observation and Patient write scopes, since both can be written
+    return if doorkeeper_authorize!(*OBSERVATION_WRITE_SCOPES)
+    return if doorkeeper_authorize!(*PATIENT_WRITE_SCOPES)
+
+    # Parse in the FHIR
+    contents = FHIR.from_contents(request.body.string)
+    errors = contents&.validate
+    status_bad_request(format_fhir_validation_errors(errors)) && return if contents.nil? || !errors.empty?
+
+    # Validate that we can go forward with processing the Bundle
+    error, path = validate_transaction_bundle(contents)
+    status_unprocessable_entity_with_custom_errors([error], path) && return unless error.nil?
+
+    # Transform all the Patients from FHIR
+    patients = []
+    contents.entry&.each_with_index do |entry, index|
+      next unless entry.resource&.resourceType&.downcase == 'patient'
+
+      resource, fhir_map = build_patient(entry&.resource)
+      change_fhir_map_context!(fhir_map, 'Patient', "Bundle.entry[#{index}].resource")
+      patients << { resource: resource, fhir_map: fhir_map, full_url: entry.fullUrl }
+    end
+
+    # Transform all of the Laboratories from FHIR
+    contents.entry&.each_with_index do |entry, index|
+      next unless entry.resource&.resourceType&.downcase == 'observation'
+
+      # We require that each Observation references a Patient in the same Bundle
+      referenced_patient = patients.find { |p| p[:full_url] == entry.resource&.subject&.reference }&.dig(:resource)
+      if referenced_patient.nil?
+        status_unprocessable_entity_with_custom_errors(
+          ['Observation resources must reference the fullUrl of a Patient in the same Bundle'],
+          "Bundle.entry[#{index}].resource.subject.reference"
+        ) && return
+      end
+
+      resource, fhir_map = build_laboratory(entry&.resource)
+      change_fhir_map_context!(fhir_map, 'Observation', "Bundle.entry[#{index}].resource")
+      referenced_patient.laboratories << resource
+      # Laboratory must be validated here since errors are inaccessible when saving Patient
+      unless resource.valid?(:api) && fhir_map.all? { |_k, v| v[:errors].blank? }
+        req_json = JSON.parse(request.body.string)
+        status_unprocessable_entity(resource, fhir_map, req_json) && return
+      end
+    end
+
+    # Save each Patient, along with its corresponding Labs
+    saved_patients = []
+    ActiveRecord::Base.transaction do
+      patients.each do |patient|
+        saved_patients << save_patient(patient[:resource], patient[:fhir_map])
```

Comment:
Yeah good catch, and an important point to think about. Since we don't do deliberate rollbacks, but instead set a status and return (either via `return` or via raising another type of error that isn't `ActiveRecord::Rollback`), I believe we are covered in this case.